drivers/bt_ll_nrfxlib/clock_control/clock_control_ll_nrfxlib.c

```diff
@@ -0,0 +1,162 @@
+{
+	ARG_UNUSED(dev);
+
+	int errcode = MULTITHREADING_LOCK_ACQUIRE();
+
+	if (errcode == 0) {
+		errcode = ble_controller_hf_clock_request(NULL);
+		MULTITHREADING_LOCK_RELEASE();
+	}
+	if (errcode != 0) {
+		return -EFAULT;
+	}
+
+	bool blocking = POINTER_TO_UINT(sub_system);
+
+	if (blocking) {
+		bool is_running = false;
+
+		while (!is_running) {
+			errcode = MULTITHREADING_LOCK_ACQUIRE();
+			if (errcode == 0) {
+				errcode = ble_controller_hf_clock_is_running(
+					&is_running);
+				MULTITHREADING_LOCK_RELEASE();
+			}
+			if (errcode != 0) {
+				return -EFAULT;
+			}
+		}
+	}
+
+	return 0;
+}
+
+static int hf_clock_stop(struct device *dev, clock_control_subsys_t sub_system)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(sub_system);
+
+	int errcode = MULTITHREADING_LOCK_ACQUIRE();
+
+	if (errcode == 0) {
+		errcode = ble_controller_hf_clock_release();
+		MULTITHREADING_LOCK_RELEASE();
+	}
+	if (errcode != 0) {
+		return -EFAULT;
+	}
+
+	return 0;
+}
+
+static int hf_clock_get_rate(struct device *dev,
+			     clock_control_subsys_t sub_system, u32_t *rate)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(sub_system);
+
+	if (rate == NULL) {
+		return -EINVAL;
+	}
+
+	*rate = MHZ(16);
+	return 0;
+}
+
+static int lf_clock_start(struct device *dev, clock_control_subsys_t sub_system)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(sub_system);
+
+	/* No-op. LFCLFK is started by default. */
+
+	return 0;
+}
+
+static int lf_clock_get_rate(struct device *dev,
+			     clock_control_subsys_t sub_system, u32_t *rate)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(sub_system);
+
+	if (rate == NULL) {
+		return -EINVAL;
+	}
+
+	*rate = 32768;
+	return 0;
+}
+
+static int clock_control_init(struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	/* No-op. Initialized by hci_driver_init() in
+	 * subsys/bluetooth/controller/hci_driver.c
+	 */
+
+	return 0;
+}
+
+static const struct clock_control_driver_api hf_clock_control_api = {
+	.on = hf_clock_start,
+	.off = hf_clock_stop,
+	.get_rate = hf_clock_get_rate,
+};
+
+DEVICE_AND_API_INIT(hf_clock,
+		    DT_NORDIC_NRF_CLOCK_0_LABEL "_16K",
+		    clock_control_init, NULL, NULL, PRE_KERNEL_1,
+		    CONFIG_KERNEL_INIT_PRIORITY_DEVICE, &hf_clock_control_api);
+
+/* LFCLK doesn't have stop function to replicate the nRF5 Power Clock driver
+ * behavior.
+ */
+static const struct clock_control_driver_api lf_clock_control_api = {
+	.on = lf_clock_start,
+	.off = NULL,
+	.get_rate = lf_clock_get_rate,
+};
+
+DEVICE_AND_API_INIT(lf_clock,
+		    DT_NORDIC_NRF_CLOCK_0_LABEL "_32K",
+		    clock_control_init, NULL, NULL, PRE_KERNEL_1,
+		    CONFIG_KERNEL_INIT_PRIORITY_DEVICE, &lf_clock_control_api);
+
+#if defined(CONFIG_USB) && defined(CONFIG_SOC_NRF52840)
+#include <nrf_power.h>
+
+void nrf5_power_usb_power_int_enable(bool enable)
+{
+	u32_t mask;
+
+
+	mask = NRF_POWER_INT_USBDETECTED_MASK |
+	       NRF_POWER_INT_USBREMOVED_MASK |
+	       NRF_POWER_INT_USBPWRRDY_MASK;
+
+	if (enable) {
+		nrf_power_int_enable(mask);
+		irq_enable(DT_NORDIC_NRF_CLOCK_0_IRQ_0);
```

Comment:
This function enables interrupts on the USB events, but the interrupt goes to ble_controller_POWER_CLOCK_IRQHandler(). Could you please confirm that the BLE stack will propagate these events to usb_dc_nrfx_power_event_callback() and clear them?